### ISSUE TITLE
Fix paging for all prompts

### ISSUE THREAD

**Issue by sassman** [mitsuhiko/dialoguer#125]
origin of this issue is https://github.com/cargo-generate/cargo-generate/issues/400 where a test example is provided and little more context is given.

**Comment by deg0nz** [comment]
Hi,

I hope, @pksunkara meant "increase usability for paging" by "Fix paging for all prompts". :)

(I looked into the code of `cargo-generate` and the error that happens in cargo-generate/cargo-generate#400, is that they use a loooooooong list without paging, which makes the select prompt unusable)

Anyways, did a quick&dirty demo with the select prompt to suggest how we could increase the usability with paging. 
Maybe that is what you had in mind...

Here's what I added:

- Prompt is always shown on every page
- A page indicator is printed next to the prompt

Here's how it looks with the select example:
![better_paging](https://user-images.githubusercontent.com/6370295/129628296-4388063d-b69e-4e0c-9c7c-5ace1397e021.gif)

I also tried it with `cargo-generate` with the repo described in cargo-generate/cargo-generate#400. It looks like this:
![better_paging_cargo_generate](https://user-images.githubusercontent.com/6370295/129631199-cb4daa70-f5d7-452e-bba4-4abcae046f46.gif)

If this is what you had in mind, I could start implementing this for all prompts that need paging and make a PR to discuss the details.

**Another thought here:**
IMHO dialoguer should figure out for itself if it needs to activate paging or not. It could just check, if `selections.len() > capacity` and activate paging automatically.

**Comment by deg0nz** [comment]
D'uh. I somehow managed to delete my comment (I can't see it anymore) 🤦‍♀️. So here we go again:

I hope, @pksunkara meant, "Increase paging usability" with "Fix paging for all prompts". :)

I looked into the code of `cargo-generate` regarding cargo-generate/cargo-generate#400. What happens there is, that they use a looooong list _without_ paging, which makes the select prompt unusable.

Anyways, I did a quick&dirty demo to suggest how we could increase the usability with paging. Here's what it could look like with the select prompt:

![better_paging](https://user-images.githubusercontent.com/6370295/129633526-1c967847-b01e-4e2d-bb94-528fd0ff50af.gif)


I also did it with the example mentioned in cargo-generate/cargo-generate#400, which then would look like this:

![better_paging_cargo_generate](https://user-images.githubusercontent.com/6370295/129633587-26b199f8-8ac6-43b8-8eed-4b9a72b8c7ac.gif)

If this is what you actually meant, I could start implementing this for all prompts that need paging and make a PR to discuss the details.

**Another thought here:**
IMHO, `dialoguer` should figure out for itself, if it needs paging or not. It could just check, if `selections.len() > term.size().0` and activate paging for itself. This would also prevent errors like cargo-generate/cargo-generate#400 in the future.

**Comment by pksunkara** [comment]
> use a looooong list without paging, which makes the select prompt unusable.

Even if paging was enabled, there are quite a lot of issues with it (especially when terminal is resized).

> IMHO, dialoguer should figure out for itself, if it needs paging or not. It could just check, if selections.len() > term.size().0 and activate paging for itself.

I agree. Basically the whole paging thing in dialoguer needs to be looked over carefully, improved and all bugs fixed. Which is what this issue is about.

**Comment by taurr** [comment]
This will be perfect, once your feature with Fuzzy Search also lands in a release :-)

**Comment by deg0nz** [comment]
Ok, good to know.

> Even if paging was enabled, there are quite a lot of issues with it (especially when terminal is resized).

@pksunkara Could you elaborate further, which issues there are? Or just throw some issue numbers at me. I would like to look into it and check, if I can fix that stuff.

**Comment by deg0nz** [comment]
I started to work on fixing paging in https://github.com/deg0nz/dialoguer/tree/Test-Paging-Stuff

I managed to extract all the paging stuff and put it into an own module without breaking the paging feature so far.

But some things came up:

- I added an `update()` function so the whole page can be re-rendered. But this only works, when a new key press occurs, because then the next `loop` iteration is started. If we really want a dynamic updating of the whole screen, we need something that observes `term.size()` output -> I think this would need changes in `console` as well and could involve concurrency and/or async stuff. But I think, that "healing" on the next keypress would suffice for now.
- The more I work on the paging stuff, the more it feels, that this should somehow be a part of `TermThemeRenderer`. Should we leave a `Paging` module, or should we integrate the paging into `TermThemeRenderer`?
